waiter/src/waiter/process_request.clj

```diff
@@ -483,6 +482,57 @@
             (meters/mark! (metrics/waiter-meter "core" "process-errors"))
             (utils/exception->response e request)))))))
 
+(defn service-id->interstitial-promise
+  "TODO (shams) docstring"
+  [service-id->interstitial-promise-atom service-id]
+  (get @service-id->interstitial-promise-atom service-id))
+
+(defn start-service-interstitial!
+  "TODO (shams) docstring"
+  [service-id->interstitial-promise-atom service-id interstitial-secs]
+  (if-let [existing-interstitial-timed-out-promise
+           (service-id->interstitial-promise service-id->interstitial-promise-atom service-id)]
+    existing-interstitial-timed-out-promise
+    (let [interstitial-timed-out-promise (promise)]
+      (->> (fn [service-id->interstitial-promise]
+             (log/info "creating interstitial promise for" service-id)
+             (if-not (contains? service-id->interstitial-promise service-id)
+               ;; TODO (shams) trigger call to start service in follow-up PR, so we can move it out of prepare instance
+               (assoc service-id->interstitial-promise service-id interstitial-timed-out-promise)
+               service-id->interstitial-promise))
+           (swap! service-id->interstitial-promise-atom))
+      ;; TODO (shams) trigger deliver as soon as we have a healthy instance for this service,
+      ;; TODO (shams) e.g. by running a separate go-block that listens to scheduler messages
```

Comment:
We might already have a healthy instance (say in the event of a Waiter restart). Are we going to wait until we have a scheduler message?

Reply:
We can work around that case by introducing a separate flag in the atom which will effectively monitor the interstitial state.